crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -3204,18 +3204,32 @@ impl<'db> TypeInferenceBuilder<'db> {
         //   f(T_inter) = f(P1) & f(P2) & ... & f(Pn)
         //
         // The reason for this is the following: In general, for any function 'f', the
-        // set f(A) & f(B) can be *larger than* the set f(A & B). This means that we
-        // will return a type that is too wide, which is not necessarily problematic.
+        // set f(A) & f(B) is *larger than or equal to* the set f(A & B). This means
+        // that we will return a type that is possibly wider than it could be, but
+        // never wrong.
         //
         // However, we do have to leave out the negative contributions. If we were to
         // add a contribution like ~f(N1), we would potentially infer result types
-        // that are too narrow, since ~f(A) can be larger than f(~A).
```

Comment:
It's better not to talk about the size of ~f(A). We should rather compare `f(A & ~B)` with `f(A) & ~f(B)`. The former can be larger (`bool`) than the latter (`Literal[True]`), as demonstrated in the example below.

Reply:
Yes, this is the part I was having the hardest time thinking through clearly

Reply:
The fully worked example makes it much clearer